File: app/src/main/java/com/teixeira/subtitles/project/ProjectManager.java
/*
 * This file is part of SubTypo.
 *
 * SubTypo is free software: you can redistribute it and/or modify it under the terms of
 * the GNU General Public License as published by the Free Software Foundation, either version 3 of
 * the License, or (at your option) any later version.
 *
 * SubTypo is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along with SubTypo.
 * If not, see <https://www.gnu.org/licenses/>.
 */

package com.teixeira.subtitles.project;

import com.teixeira.subtitles.models.Project;

/**
 * Class to manipulate project data.
 *
 * @author Felipe Teixeira
 */
public class ProjectManager {

  private static volatile ProjectManager sInstance;

  public static ProjectManager getInstance() {
    if (sInstance == null) {
      synchronized (ProjectManager.class) {
        if (sInstance == null) sInstance = new ProjectManager();
      }
    }
    return sInstance;
  }

  private Project project = null;

  private ProjectManager() {}

  /**
   * Configure and open the provided project.
   *
   * @param The project to open.
   */
  public void openProject(Project project) {
    this.project = project;
  }

  /**
   * Get Open Project Instance.
   *
   * @return The open project instance.
   */
  public Project getProject() {
    return this.project;
  }

  /** Destroy and close the open project. */
  public void destroy() {
    this.project = null;
  }
}

File: app/src/main/java/com/teixeira0x/subtypo/project/ProjectManager.kt
package com.teixeira0x.subtypo.project

import androidx.media3.common.Format
import com.teixeira0x.subtypo.models.Project
import com.teixeira0x.subtypo.project.internal.ProjectManagerImpl

interface ProjectManager {

  companion object {

    private var sInstance: ProjectManager? = null

    @JvmStatic
    fun getInstance(): ProjectManager {
      return sInstance
        ?: synchronized(this) {
          sInstance ?: ProjectManagerImpl().also { sInstance = it }
        }
    }
  }

  /** Open project */
  val project: Project?

  /** Open project video format */
  val videoFormat: Format?

  /**
   * Configure and open the provided project.
   *
   * @param The project to open.
   */
  fun openProject(project: Project)

  /**
   * Update current video information.
   *
   * @param videoFormat The player's video format.
   */
  fun updateVideoFormat(videoFormat: Format?)

  /** destroy */
  fun destroy()
}
